packages/zoe/src/z2spec/escrow-exo.js

```diff
@@ -0,0 +1,311 @@
+        },
+      },
+    },
+  );
+
+const assertKeywordSubset = (partLabel, part, whole) => {
+  const wholeKeys = keys(whole);
+  const extraKeys = keys(part).filter(k => !wholeKeys.includes(k));
+  extraKeys.length === 0 ||
+    Fail`unexpected ${bare(partLabel)} keywords: ${q(extraKeys)} not in ${q(wholeKeys)}`;
+};
+
+/**
+ *  @param {Zone} zone
+ * @param {ZCF['makeInvitation']} makeInvitation
+ */
+export const prepareEscrowExchange = (zone, makeInvitation) => {
+  const makeDurablePublishKit = prepareDurablePublishKit(
+    zone.mapStore('publish baggage'),
+    'escrow publisher',
+  );
+  const makeSeatKit = prepareSeatKit(zone, makeDurablePublishKit);
+
+  const makeOwnableSeat = prepareOwnable(
+    zone.subZone('ownableSeat'),
+    makeInvitation,
+    'OwnableSeat',
+    /* XXX @type {(keyof EscrowSeat)[]} */
+    /** @type {const} */ ([
+      'readOnly',
+      'hasExited',
+      'getProposal',
+      'getCurrentAllocation',
+      'getExitSubscriber',
+      'getPayout',
+      'getPayouts',
+      'exit',
+      'fail',
+    ]),
+  );
+
+  const makeOwnableSeatResolver = prepareOwnable(
+    zone.subZone('ownableSeatResolver'),
+    makeInvitation,
+    'OwnableSeatResolver',
+    /* XXX @type {(keyof EscrowSeatResolver)[]} */
+    /** @type {const} */ (['readOnly', 'exit', 'fail', 'updateAllocation']),
+  );
+
+  const makeEscrowExchange = zone.exoClass(
+    'EscrowExchange',
+    M.interface('EscrowExchangeI', {}, { defaultGuards: 'passable' }),
+    () => ({
+      /** @type {IssuerKeywordRecord} */
+      issuers: {},
+      /** @type {Record<Keyword, Purse>} */
+      escrowPurses: {},
+    }),
+    {
+      /**
+       * TODO: use brands instead of keywords?
+       *
+       * @param {Keyword} keyword
+       * @param {Issuer} issuer
+       */
+      async addIssuer(keyword, issuer) {
+        !(keyword in this.state.issuers) ||
+          Fail`keyword alread used: ${keyword}`;
+        const purse = await E(issuer).makeEmptyPurse();
+        const { issuers, escrowPurses } = this.state;
+        !(keyword in issuers) || Fail`keyword alread used: ${keyword}`;
+        this.state.issuers = harden({ ...issuers, [keyword]: issuer });
+        this.state.escrowPurses = harden({ ...escrowPurses, [keyword]: purse });
+      },
+      getIssuers() {
+        return this.state.issuers;
+      },
+      /**
+       * TODO: with the escrow service separate from the contract host,
+       * is the exit rule relevant to the escrow service?
+       *
+       * @param {ProposalRecord} proposal
+       * @param {PaymentKeywordRecord} payments
+       */
+      async makeEscrowSeatKit(proposal, payments) {
+        const { escrowPurses } = this.state;
+
+        assertKeywordSubset('give', proposal.give, escrowPurses);
+        assertKeywordSubset('want', proposal.want, escrowPurses);
+        assertKeywordSubset('payment', payments, escrowPurses);
+        const giveKeys = keys(proposal.give);
+
+        // TODO: partial failure
```

Comment:
This is the issue of a misbehavior issuer blocking assets from well behaved issuers, right?

https://github.com/Agoric/agoric-sdk/issues/1271